code/datums/components/scope.dm

```diff
@@ -0,0 +1,274 @@
+/datum/component/scope
+	/// How far we can extend, with modifier of 1, up to our vision edge, higher numbers multiply.
+	var/range_modifier = 1
+	/// Fullscreen object we use for tracking.
+	var/obj/screen/fullscreen/cursor_catcher/scope/tracker
+	/// The owner of the tracker's ckey. For comparing with the current owner mob, in case the client has left it (e.g. ghosted).
+	var/tracker_owner_ckey
+	/// The method which we zoom in and out
+	var/zoom_method = ZOOM_METHOD_ITEM_ACTION
+	/// if not null, an item action will be added. Redundant if the mode is ZOOM_METHOD_RIGHT_CLICK or ZOOM_METHOD_WIELD.
+	var/item_action_type
+	/// Time to scope up, if you want the scope to take time to boot up. Used by the LWAP
+	var/time_to_scope
+	/// Are we currently trying to scope in?
+	var/attempting_to_scope = FALSE
+	/// Do we let the user scope and click on the middle of their screen?
+	var/allow_middle_click = FALSE
+
+/datum/component/scope/Initialize(range_modifier = 1, zoom_method = ZOOM_METHOD_ITEM_ACTION, item_action_type = /datum/action/zoom, time_to_scope = 0, allow_middle_click = FALSE)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.range_modifier = range_modifier
+	src.zoom_method = zoom_method
+	src.item_action_type = item_action_type
+	src.time_to_scope = time_to_scope
+	src.allow_middle_click = allow_middle_click
+
+
+/datum/component/scope/Destroy(force)
+	if(tracker)
+		stop_zooming(tracker.owner)
+	return ..()
+
+/datum/component/scope/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, PROC_REF(on_move))
+	switch(zoom_method)
+		if(ZOOM_METHOD_WIELD)
+			RegisterSignal(parent, SIGNAL_ADDTRAIT(TRAIT_WIELDED), PROC_REF(on_wielded))
+			RegisterSignal(parent, SIGNAL_REMOVETRAIT(TRAIT_WIELDED), PROC_REF(on_unwielded))
+	if(item_action_type)
+		var/obj/item/parent_item = parent
+		var/datum/action/scope = new item_action_type(parent)
+		parent_item.actions += scope
+		RegisterSignal(scope, COMSIG_ACTION_TRIGGER, PROC_REF(on_action_trigger))
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+	if(istype(parent, /obj/item/gun))
+		RegisterSignal(parent, COMSIG_GUN_TRY_FIRE, PROC_REF(on_gun_fire))
+
+/datum/component/scope/UnregisterFromParent()
+	if(item_action_type)
+		var/obj/item/parent_item = parent
+		var/datum/action/scope = locate(item_action_type) in parent_item.actions
+		parent_item.actions -= (scope)
+	UnregisterSignal(parent, list(
+		COMSIG_MOVABLE_MOVED,
+		SIGNAL_ADDTRAIT(TRAIT_WIELDED),
+		SIGNAL_REMOVETRAIT(TRAIT_WIELDED),
+		COMSIG_GUN_TRY_FIRE,
+		COMSIG_PARENT_EXAMINE,
+	))
+
+/datum/component/scope/process()
+	var/mob/user_mob = tracker.owner
+	var/client/user_client = user_mob.client
+	if(!user_client)
+		stop_zooming(user_mob)
+		return
+	tracker.calculate_params()
+	animate(user_client, world.tick_lag, pixel_x = tracker.given_x, pixel_y = tracker.given_y)
+
+/datum/component/scope/proc/on_move(atom/movable/source, atom/oldloc, dir, forced)
+	SIGNAL_HANDLER
```

Comment:
Huh. not seen that on many signal_handlers I have worked with?

Reply:
@lewcc suggested it to me here: https://github.com/ParadiseSS13/Paradise/pull/24626#discussion_r1525194827

Perhaps I'm reading/remembering too far into it, but it seems like a good idea to start doing that everywhere, for ease of searching as I said. We've already used it in a few places (https://github.com/search?q=repo%3AParadiseSS13%2FParadise+%2FSIGNAL_HANDLER+*%5C%2F%2F&type=code), Might as well spread it, and increase the coverage.

Reply:
Yeah, that's mostly just been a little thing I've started doing myself to help keep track of the signals my callbacks referred to. I think it would be nice if we adopted it, since I feel like callbacks can sometimes require digging to find their associated un/register signal, but it's by no means mandatory at this point

Reply:
done